lib/core/buffer.cpp

```diff
@@ -34,45 +40,25 @@ struct zero_frames_thread_args {
     int ID;
 };
 
-void* private_zero_frames(void* args);
-
-// Returns -1 if there is no consumer with that name
-int private_get_consumer_id(Buffer* buf, const char* name);
-
-// Returns -1 if there is no producer with that name
-int private_get_producer_id(Buffer* buf, const char* name);
-
-// Marks the consumer named by `name` as done for the given ID
-void private_mark_consumer_done(Buffer* buf, const char* name, const int ID);
-
-// Marks the producer named by `name` as done for the given ID
-void private_mark_producer_done(Buffer* buf, const char* name, const int ID);
-
-// Returns 1 if all consumers are done for the given ID.
-int private_consumers_done(Buffer* buf, const int ID);
+typedef std::lock_guard<std::recursive_mutex> buffer_lock;
 
-// Returns 1 if all producers are done for the given ID.
-int private_producers_done(Buffer* buf, const int ID);
+GenericBuffer::GenericBuffer(const std::string& _buffer_name, const std::string& _buffer_type,
+                             metadataPool* pool, int _num_frames) :
+    num_frames(_num_frames),
+    shutdown_signal(false), buffer_name(_buffer_name), buffer_type(_buffer_type),
+    metadata_pool(pool), metadata(num_frames, nullptr) {}
```

Comment:
I did set the log prefix and level in `bufferFactory` -- would you prefer to do it in constructors?

Reply:
Oh I missed that somehow!  I don't have a strong preference.  Slightly nicer to have it in the constructor since then the factory doesn't need to remember to do that, and it's a bit cleaner, but realistically that's the only time this will happen.  Eventually it would be nice to move to an abstract factory for the buffers as well, but even then, only the factory needs to remember to include this logic and it's the same for every type.

Reply:
The log level comes from the config, which currently isn't passed to the Buffer constructor (maybe it should be...), so I kept the set_log_level() in the bufferFactory, but moved the set_log_prefix to the GenericBuffer constructor.